lib/ansible/galaxy/role.py

```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
                 temp_file.close()
                 return temp_file.name
             except Exception as e:
-                display.error(u"failed to download the file: %s" % to_text(e))
+                display.error("failed to download the file: %s" % to_text(e))
 
         return False
 
+    @staticmethod
+    def _is_path_sane(path: str) -> bool:
+        """
+        Check if the given path from a tar member is safe to extract.
+
+        This method looks for common risks that can lure in the path:
+        - Absolute paths that allow to create files outside of the target.
+        - `..` parts that allow to write files outside of the target.
+
+        It does NOT check:
+        - Symbols that are not allowed in environments (like `|` in Windows).
+        - Embedded variables like `$HOME/.config`. Please do not pass paths to a shell.
+        """
+
+        if path.startswith('/'):
+            return False
+        if [True for part in path.split(os.sep) if part == ".."]:
+            return False
+        return True
+
+    def _extract(self, role_tar_file, archive_parent_dir, path):
+        # we strip off any higher-level directories for all of the files contained within
+        # the tar file here. The default is 'github_repo-target'. Gerrit instances, on the other
+        # hand, does not have a parent directory at all.
+        display.display("- extracting %s to %s" % (self.name, self.path))
+
+        if not os.path.exists(self.path):
+            os.makedirs(self.path)
+        elif not os.path.isdir(self.path):
+            raise AnsibleError("the specified roles path exists and is not a directory.")
+        elif not context.CLIARGS.get("force", False):
+            raise AnsibleError("the specified role %s appears to already exist. Use --force to replace it." % self.name)
+        else:
+            # using --force, remove the old path
+            if not self.remove():
+                raise AnsibleError("%s doesn't appear to contain a role.\n  please remove this directory manually if you "
+                                   "really want to put the role here." % self.path)
+
+        try:
+            # now we do the actual extraction to the path
+            for member in role_tar_file.getmembers():
+                # Strip parent path from tar members (`rolename/tasks/main.yml` becomes `tasks/main.yml`).
+                member.name = member.name.replace(archive_parent_dir + os.sep, "", 1)
+
+                # Only extract regular files and symbolic links.
+                if not member.isreg() and not member.issym():
+                    continue
+
+                if not self._is_path_sane(member.name) and not context.CLIARGS.get("force", False):
```

Comment:
The method name with "sane" suggests quite a broad meaning. It's better to be more specific about what that method does. Also, does it need to be a method? I'd have a pure function instead.